ports/nrf/common-hal/alarm/__init__.c

```diff
@@ -0,0 +1,399 @@
+    sleepmem_wakeup_pin   = WAKEUP_PIN_UNDEF;
+
+#ifdef NRF_DEBUG_PRINT
+    int ct = 40;
+    char reason = '?';
+#define WAKEUP_REASON(x) reason = (x)
+#else
+#define WAKEUP_REASON(x)
+#endif
+
+    while(1) {
+	if (mp_hal_is_interrupted()) {
+	    WAKEUP_REASON('I');
+	    break;
+	}
+        if (serial_connected() && serial_bytes_available()) {
+	    WAKEUP_REASON('S');
+	    break;
+	}
+        RUN_BACKGROUND_TASKS;
+	wakeup_cause = alarm_woken_from_sleep_2();
+	if (wakeup_cause != NRF_SLEEP_WAKEUP_UNDEFINED) {
+	    WAKEUP_REASON('0'+wakeup_cause);
+	    break;
+	}
+	if (have_timeout) {
+	    remaining = end_tick - port_get_raw_ticks(NULL);
+	    // We break a bit early so we don't risk setting the alarm before the time when we call
+	    // sleep.
+	    if (remaining < 1) {
+	        WAKEUP_REASON('t');
+	        break;
+	    }
+	    port_interrupt_after_ticks(remaining);
+	}
+        // Idle until an interrupt happens.
+        port_idle_until_interrupt();
+#ifdef NRF_DEBUG_PRINT
+        if (ct > 0) {
+            dbg_printf("_");
+            --ct;
+        }
+#endif
+	if (have_timeout) {
+	    remaining = end_tick - port_get_raw_ticks(NULL);
+	    if (remaining <= 0) {
+	        wakeup_cause = NRF_SLEEP_WAKEUP_TIMER;
+		sleepmem_wakeup_event = SLEEPMEM_WAKEUP_BY_TIMER;
+	        WAKEUP_REASON('T');
+		break;
+	    }
+	}
+    }
+#ifdef NRF_DEBUG_PRINT
+    dbg_printf("%c\r\n", reason);
+#endif
+
+#if defined(MICROPY_QSPI_CS)
+    qspi_flash_exit_sleep();
+#endif
+
+    tickdiff = port_get_raw_ticks(NULL) - start_tick;
+    if (prescaler == 0) {
+        timediff_ms = tickdiff / 1024;
+    }
+    else {
+        timediff_ms = tickdiff * prescaler / 1024;
+    }
+    (void)timediff_ms;
+#ifdef NRF_DEBUG_PRINT
+    dbg_printf("lapse %6.1f sec\r\n", (double)(timediff_ms));
+#endif
+
+    return wakeup_cause;
+}
+
+mp_obj_t common_hal_alarm_light_sleep_until_alarms(size_t n_alarms, const mp_obj_t *alarms) {
+    mp_obj_t r_obj = mp_const_none;
+    alarm_time_timealarm_clear_wakeup_time();
+    _setup_sleep_alarms(false, n_alarms, alarms);
+
+#ifdef NRF_DEBUG_PRINT
+    dbg_printf("\r\nlight sleep...");
+#endif
+
+    int64_t timediff_ms = alarm_time_timealarm_get_wakeup_timediff_ms();
+    nrf_sleep_source_t cause = system_on_idle_until_alarm(timediff_ms, 0);
+    (void)cause;
+
+#ifdef NRF_DEBUG_PRINT
+    //dbg_printf("wakeup! ");
+    print_wakeup_cause(cause);
+#endif
+
+    if (mp_hal_is_interrupted()) {
+#ifdef NRF_DEBUG_PRINT
+        dbg_printf("mp_hal_is_interrupted\r\n");
+#endif
+        r_obj = mp_const_none;
+    }
+    else {
+        r_obj = _get_wake_alarm(n_alarms, alarms);
+    }
+    alarm_reset();
+    return r_obj;
+}
+
+void common_hal_alarm_set_deep_sleep_alarms(size_t n_alarms, const mp_obj_t *alarms) {
+    alarm_time_timealarm_clear_wakeup_time();
+    _setup_sleep_alarms(true, n_alarms, alarms);
+}
+
+#define PRESCALER_VALUE_IN_DEEP_SLEEP (1024)
+
+void NORETURN alarm_enter_deep_sleep(void) {
```

Comment:
How does this handle interaction with the mandatory 5s delay (CIRCUITPY_USB_CONNECTED_SLEEP_DELAY) after first startup? That relies on the use of `port_interrupt_after_ticks()` to exit when it is ready to enter true deep sleep. It caused problems on my STM32 implementation if the sleep duration was less than five seconds - you implementation may catch it automatically but I figured I'd check in about it.

Reply:
@hierophect I did not change anything around the 5s delay check, but I think true deep sleep may cause the delay because RTC is prescaled and port_get_raw_ticks() would return a wrong value. This may occur when your .py code try to enter true deep sleep too soon (<5secs) after wake up from the last sleep.
Maintaining tick value even when the prescaler is enabled is quite complicated...

Reply:
I see now that your system works a little different, in that `alarm_time_timealarm_set_alarms` does not set an interrupt.